articles/102_dependency_groups.md

```diff
@@ -0,0 +1,164 @@
+For the `satisfy-at-least-one` semantic the release tool could either use a platform specific mechanism like `provides` in Debian control files or simply expand the group to all packages (same as `satisfy-all`).
+This design document doesn't aim to specify what the "best" behavior for the release tool is since the decision is likely platform dependent.
+
+Since the release tool expands the dependency groups when being invoked, packages which join the group after that release won't be used until the package defining the group dependency is being re-released.
+
+### rosinstall_generator
+
+The `rosinstall_generator` already utilized all manifests from the `rosdistro` to decide which packages / repositories should be fetched.
+This design document doesn't aim to specify what the "best" behavior for this tool is.
+It could either match the behavior of the build tool or of the release tool or even something different.
+
+## Information in the Manifest
+
+To provide the necessary metadata for the build and release tools the following information must be provided by the manifest:
+
+* (A) A specific package declares a group dependency and uses a name to identify the group.
+  The dependency can be of any type specified in [REP 140](http://www.ros.org/reps/rep-0140.html): e.g. `build`, `exec`, `test`, `doc`, etc.
+
+* (B) The semantic of the group dependency needs to be defined (in case (1) `satisfy-at-least-one`, in case (2) `satify-all`).
+
+* (C) A list of preferred / excluded packages can be enumerated for each dependency group.
+
+* (D) Any package can declare that it is part of a group by name.
+
+### Possible Format Extensions
+
+The group dependency (A) could be specified in two ways:
+
+* (A.1) Using the existing `*_depend` tags and adding an attribute to the tag to make it a group dependency.
+* (A.2) Defining new `*_group_depend` tags which specifically identify group dependencies.
+
+Considering the existing API in `catkin_pkg` which exposes the different dependencies as members of the `Package` class (A.1) would require existing code to distinguish the type of the dependency (group vs. no-group).
+On the other hand (A.2) will simply add additional members which doesn't require existing code to change until it wants to support handling group dependencies.
+
+The semantic of the group dependency (B) can be modeled as an attribute of the group dependency tag.
+
+Each preferred / excluded package (C) name can be listed in child-tags of the dependency group to keep the information atomic.
+
+The group membership (D) could be specified in two ways:
+
+* (D.1) Adding a new tag to the `export` section: e.g. `member_of_group`.
+* (D.2) Defining a new tag in the `package` section: e.g. `member_of_group`.
+
+#### Example for use case (1)
+
+A package declaring a group dependency:
+
+```
+<package>
+  <name>rmw_implementation</name>
+  <group_depend satisfy="at-least-one">i_am_a_rmw_implementation</group_depend>
```

Comment:
Since `at-least-one` is pretty verbose I was considering `any` (based on the [Python function](https://docs.python.org/3/library/functions.html#any)). Does that sound reasonable to native speakers?

Reply:
Personally, I like `any` better.

Reply:
'any' and 'all' would make sense. The other thing not discussed here would be 'one', where it would require one and only one from the group, but we can also scope that out. That's sometimes done where there's multiple implementations but they're not side by side installable.

Reply:
Yeah, `any` is probably fine. I originally advised for the more specific key phrase, as I felt any could mean zero as well, e.g. in regex `*` is any and `+` is at least one.